src/datumaro/components/algorithms/hash_key_inference/explorer.py

```diff
@@ -74,6 +74,32 @@ def explore_topk(
 
         database_keys = self._database_keys
 
+        if isinstance(query, list):
+            topk_for_query = int(topk // len(query)) * 2 if not len(query) == 1 else topk
+            query_hash_key_list = []
+            result_list = []
+            for query_ in query:
+                if isinstance(query_, DatasetItem):
+                    query_key = self._get_hash_key_from_item_query(query_)
+                    query_hash_key_list.append(query_key)
+                elif isinstance(query_, str):
+                    query_key = self._get_hash_key_from_text_query(query_)
+                    query_hash_key_list.append(query_key)
+                else:
+                    raise MediaTypeError(
+                        "Unexpected media type of query '%s'. "
+                        "Expected 'DatasetItem' or 'string', actual'%s'" % (query_, type(query_))
+                    )
+
+            for query_key in query_hash_key_list:
+                unpacked_key = np.unpackbits(query_key.hash_key, axis=-1)
+                logits = calculate_hamming(unpacked_key, database_keys)
+                ind = np.argsort(logits)
+
+                item_list = np.array(self._item_list)[ind]
+                result_list.extend(item_list[:topk_for_query].tolist())
+            return np.random.choice(result_list, topk)
```

Comment:
Why is there a non-deterministic process here? Last time you said that the N-to-M top-k similarity search 1) finds the k similarity items for N queries (Nk candidates) and 2) pick up the top-k items from them, is it right?

Reply:
I updated the process for this. First make N*topk or N*topk_for_query candidates and logits for N queries and sort logits by value. Resort candidates through logit indices and pick up top-k items from them.